libcamera-sys/c_api/controls.cpp

```diff
@@ -13,10 +13,87 @@ const char *libcamera_control_name(libcamera_control_id_t *control){
     return control->name().c_str();
 }
 
-enum libcamera_control_type libcamera_control_type(libcamera_control_id_t *control) {
+enum libcamera_control_type libcamera_control_id_type(libcamera_control_id_t *control) {
     return (enum libcamera_control_type) control->type();
 }
 
+const char *libcamera_control_id_vendor(libcamera_control_id_t *control) {
+    return control->vendor().c_str();
+}
+
+enum libcamera_control_direction libcamera_control_id_direction(libcamera_control_id_t *control) {
+    using Underlying = std::underlying_type_t<libcamera::ControlId::Direction>;
+    Underlying bits = static_cast<Underlying>(control->direction());
+
+    // …then cast that integer into your C enum
+    return static_cast<enum libcamera_control_direction>(bits);
+}
+
+bool libcamera_control_id_is_input(libcamera_control_id_t *control) {
+    return control->isInput();
+}
+
+bool libcamera_control_id_is_output(libcamera_control_id_t *control) {
+    return control->isOutput();
+}
+
+bool libcamera_control_id_is_array(libcamera_control_id_t *control) {
+    return control->isArray();
+}
+
+size_t libcamera_control_id_size(libcamera_control_id_t *control) {
+    return control->size();
+}
+
+size_t libcamera_control_id_enumerator_count(libcamera_control_id_t *control) {
+    return control->enumerators().size();
+}
+
+const char *libcamera_control_id_enumerator_name(libcamera_control_id_t *control, int32_t value) {
+    auto &rev = control->enumerators();
+    auto it = rev.find(value);
+    return it != rev.end() ? it->second.c_str() : nullptr;
+}
+
+int32_t libcamera_control_id_enumerator_value(libcamera_control_id_t *control, const char *name) {
+    if (!control || !name)
+        return 0;
+    for (auto &p : control->enumerators()) {
+        if (p.second == name)
+            return p.first;
+    }
+    return 0;
+}
+
+size_t libcamera_control_id_enumerators_len(libcamera_control_id_t *ctrl) {
+    if (!ctrl)
```

Comment:
I'm not sure if we need null check in C shim when we can ensure this by rust type system without any performance cost. For example it should not be possible to have an invalid control id, because ControlId is enum with fixed values, so ControlId::as_ptr() is never null. Maybe add an assert from the rust side instead?